sklearn/tree/_tree.pyx

```diff
@@ -291,16 +293,26 @@ cdef class DepthFirstTreeBuilder(TreeBuilder):
 
 # Best first builder ----------------------------------------------------------
 
-cdef inline int _add_to_frontier(PriorityHeapRecord* rec,
-                                 PriorityHeap frontier) nogil except -1:
-    """Adds record ``rec`` to the priority queue ``frontier``
+cdef struct PriorityHeapRecord:
+    SIZE_t node_id
+    SIZE_t start
+    SIZE_t end
+    SIZE_t pos
+    SIZE_t depth
+    bint is_leaf
+    double impurity
+    double impurity_left
+    double impurity_right
+    double improvement
```

Comment:
I think `PriorityQueue` makes sense, but as a heap maintains an order for a priority, isn't `PriorityHeap` a pleonasm?

Can this data-structure be renamed to convey the information it stores for the algorithm?

For instance could this be a `NodeRecord` or `FrontierRecord`?
```suggestion
cdef struct NodeRecord:
    """Record of information of a Node, the frontier for a split.
    
    Those records are maintained in a heap to access the Node
    with the best improvement in impurity, allowing growing
    trees greedily on this improvement.
    """
    SIZE_t node_id
    SIZE_t start
    SIZE_t end
    SIZE_t pos
    SIZE_t depth
    bint is_leaf
    double impurity
    double impurity_left
    double impurity_right
    double improvement
```

Reply:
Turns out docstrings in a `cdef struct` leads to a compiler error. I updated the PR to use a code comment instead.